src/containers/barray.mli

```diff
@@ -0,0 +1,213 @@
+    The current implementation uses Merge Sort. It uses [n/2]
+    words of heap space, where [n] is the length of the t.
+    It is usually faster than the current implementation of {!Barray.sort}.
+*)
+
+val fastSort : ('a -> 'a -> int) -> 'a t -> unit
+(** Same as {!Barray.sort} or {!Barray.stableSort}, whichever is faster
+    on typical input.
+*)
+
+val sorted : ('a Comparator.t) -> 'a t -> 'a t
+(** [sorted cmp a] makes a copy of [a] and sorts it with [cmp]. *)
+
+val sortIndices : ('a Comparator.t) -> 'a t -> int t
+(** [sortIndices cmp a] returns a new t [b], with the same length as [a],
+    such that [b.(i)] is the index of the [i]-th element of [a] in [sort cmp a].
+    In other words, [map (fun i -> a.(i)) (sortIndices a) = sorted cmp a].
+    [a] is not modified. *)
+
+val sortRanking : ('a Comparator.t) -> 'a t -> int t
+(** [sortRanking cmp a] returns a new t [b], with the same length as [a],
+    such that [b.(i)] is the position in [sorted cmp a] of the [i]-th
+    element of [a].
+    [a] is not modified.
+
+    In other words, [map (fun i -> (sorted cmp a).(i)) (sortRanking cmp a) = a].
+*)
+
+val find : ('a -> 'b option) -> 'a t -> 'b option
+
+val findWithIndex : (int -> 'a -> 'b option) -> 'a t -> 'b option
+(** Like {!find}, but also pass the index to the predicate function. *)
+
+val findIndex : ('a -> bool) -> 'a t -> (int * 'a) option
+(** [Barray.findIndex p x] returns [Some (i,x)] where [x] is the [i]-th element of [l],
+    and [p x] holds. Otherwise returns [None] *)
+
+val bsearch : 'a Comparator.t -> 'a -> 'a t ->
+  [ `All_lower | `All_bigger | `Just_after of int | `Empty | `At of int ]
+(** [Barray.bsearch cmp x arr] finds the index of the object [x] in the t [arr],
+    provided [arr] is {b sorted} using [cmp]. If the t is not sorted,
+    the result is not specified (may raise Invalid_argument).
+
+    Complexity: O(log n) where n is the length of the t
+    (dichotomic search).
+
+    @return
+    - [`At i] if [cmp arr.(i) x = 0] (for some i)
+    - [`All_lower] if all elements of [arr] are lower than [x]
+    - [`All_bigger] if all elements of [arr] are bigger than [x]
+    - [`Just_after i] if [arr.(i) < x < arr.(i+1)]
+    - [`Empty] if the t is empty *)
+
+val forAll: ('a -> bool) -> 'a t -> bool
+
+val exists: ('a -> bool) -> 'a t -> bool
+
+val count: ('a -> bool) -> 'a t -> int
+
+val iter : ('a -> unit) -> 'a t -> unit
+
+val shuffle : 'a t -> unit
+(** Shuffle randomly the t, in place *)
+
+val shuffleWith : Random.State.t -> 'a t -> unit
+(** Like shuffle but using a specialized random state *)
+
+val filter : ('a -> bool) -> 'a t -> 'a t
+(** Filter elements out of the t. Only the elements satisfying
+    the given predicate will be kept. *)
+
+val filterMap : ('a -> 'b option) -> 'a t -> 'b t
+(** Map each element into another value, or discard it *)
+
+val flatMap : ('a -> 'b t) -> 'a t -> 'b t
+(** Transform each element into an t, then flatten *)
+
+val (--) : int -> int -> int t
```

Comment:
Should we put operators in 'Infix' or 'Ops' sub modules. The motivation for this is that operators are only useful when the module they are defined is opened. However opening a module is usually a bad practice so in order to limit the scope of what is open operator can be placed in a sub module.